### PR TITLE
Add update StateValue and WriteOp to support metadata tracking

### PR DESCRIPTION
**Pull request by msmouse**
…king

* StateValue now has a new flavor which carries StateValueMetadata.
* similarly there is now WriteOp::CreationWithMetadata and friends

### Test Plan

replay-verify: https://github.com/aptos-labs/aptos-core/actions/runs/4299055448

### CODE REVIEW

**Review by zekun000 — APPROVED** [review]

File: types/src/write_set.rs [hunk 1]
```diff
@@ -19,42 +22,161 @@ pub enum WriteOp {
...
-            WriteOp::Creation(_) | WriteOp::Deletion => false,
+            WriteOp::Modification(_) | WriteOp::ModificationWithMetadata { .. } => true,
+            WriteOp::Creation(_)
+            | WriteOp::CreationWithMetadata { .. }
+            | WriteOp::Deletion
+            | WriteOp::DeletionWithMetadata { .. } => false,
+        }
+    }
+
+    pub fn squash(op: &mut Self, other: Self) -> Result<bool> {
```
Comment (grao1991) [inline types/src/write_set.rs:72]: nit: document the meaning of the returned bool value

File: types/src/write_set.rs [hunk 2]
```diff
@@ -19,42 +22,161 @@ pub enum WriteOp {
...
+        use WriteOp::*;
+
+        // n.b. With write sets from multiple sessions being squashed together, it's possible
+        //   to see two ops carrying different metadata (or one with it the other without)
+        //   due to deleting in one session and recreating in another. The original metadata
+        //   shouldn't change due to the squash.
```
Comment (grao1991) [inline types/src/write_set.rs:78]: Does this mean I cannot change the billing account of an item in a transaction?

Comment (msmouse) [reply types/src/write_set.rs:78]: Right, you can't. The write set is the end result of the entire txn, which can express "a thing is modified" but not "a thing is deallocated and then re-created" in the first place.

File: types/src/write_set.rs (hunk 1 above)
Comment (msmouse) [reply types/src/write_set.rs:76]: added

**Review by grao1991 — APPROVED** [review]

**Comment by lightmark** [comment]
Is this for refund?

**Pull request merged**